### PR TITLE
Feature/can filter get jobs with optional label argument/#64

### PR DESCRIPTION
**Pull request by oliver-tarrant-tessella**
Add the ability to filter jobs by their `userLabel`. An unfiltered list is still returned by providing an empty body  or by using `"label": "ALL"` as the request body. When submitting a job the users are no longer allowed to use the argument `label: "ALL` (case insensitive) to avoid clashes. 

## To test:

* System and unit tests

## Stories covered:

#64

### CODE REVIEW

File: SystemTests/roles/robot/files/piezo.robot [hunk 1]
```diff
@@ -170,9 +170,29 @@ Get List Of SparkApplications Includes Submitted Jobs
     ${job_name_2}=    Get Response Job Name   ${response2}
     ${job_name_3}=    Get Response Job Name   ${response3}
     Sleep     5 seconds
-    ${request_response}=    Get List Of Spark Jobs
+    ${body}=    Create Dictionary     label=ALL
+    ${request_response}=    Get List Of Spark Jobs    ${body}
```
Comment (robert-clegg-tessella) [inline SystemTests/roles/robot/files/piezo.robot]: Is there still a system test where no `label` argument is given?

File: piezo_web_app/PiezoWebApp/src/services/spark_job/spark_job_service.py [hunk 2]
```diff
@@ -81,6 +81,11 @@ def get_jobs(self):
             spark_jobs = {
                 item['metadata']['name']: SparkJobService._retrieve_status(item) for item in api_response['items']
             }
+            if label.upper() != "ALL":
+                for item in api_response['items']:
+                    user_label = SparkJobService._get_user_label(item)
+                    if user_label != label:
+                        del spark_jobs[item['metadata']['name']]
```
Comment (robert-clegg-tessella) [inline piezo_web_app/PiezoWebApp/src/services/spark_job/spark_job_service.py]: Is it not possible to give the required label when calling the kubernetes adapter?

File: piezo_web_app/PiezoWebApp/src/services/spark_job/validation/argument_validator.py [hunk 3]
```diff
@@ -20,6 +22,16 @@ def validate(key, value, validation_rule):
     raise ValueError(f"Unexpected argument {key}")
 
 
+def _validate_label(value):
+    if not isinstance(value, str):
+        return ValidationResult(False, '"label" input must be a string', None)
+    if is_str_empty(value):
+        return ValidationResult(False, '"label" input cannot be empty', None)
+    if value.upper() == "ALL":
+        return ValidationResult(False, '"label" input cannot be "ALL"', None)
```
Comment (robert-clegg-tessella) [inline piezo_web_app/PiezoWebApp/src/services/spark_job/validation/argument_validator.py]: If the user has input `"all"` then this message won't make complete sense. Add `(case insensitive)` to the end of the message.

File: piezo_web_app/PiezoWebApp/tests/handlers/get_jobs_test.py [hunk 4]
```diff
@@ -63,3 +63,28 @@ def test_get_returns_empty_array_when_no_jobs_present(self):
...
+        self.mock_spark_job_service.get_jobs.return_value = {
+            "message": "Found 2 spark jobs",
+            "spark_jobs": {"job_1": "COMPLETED", "job_3": "PENDING"},
+            "status": 200
+        }
+        body = {"label": "test-label"}
+        # Act
+        response_body, response_code = yield self.send_request(body)
+        # Assert
+        self.mock_spark_job_service.get_jobs.assert_called_once()
```
Comment (robert-clegg-tessella) [inline piezo_web_app/PiezoWebApp/tests/handlers/get_jobs_test.py]: `assert_called_once_with(...)`

File: piezo_web_app/PiezoWebApp/tests/services/spark_job/spark_job_service_test.py [hunk 5]
```diff
@@ -77,7 +77,7 @@ def test_get_jobs_sends_expected_arguments(self):
         # Arrange
         self.mock_kubernetes_adapter.list_namespaced_custom_object.return_value = {"items": []}
         # Act
-        result = self.test_service.get_jobs()
```
Comment (robert-clegg-tessella) [inline piezo_web_app/PiezoWebApp/tests/services/spark_job/spark_job_service_test.py]: We need a test showing what happens when `label` is `None`

File: piezo_web_app/PiezoWebApp/tests/services/spark_job/validation/argument_validator_test.py [hunk 6]
```diff
@@ -4,6 +4,26 @@
...
+    # Arrange
+    validation_rule = ValidationRule({'classification': 'Optional'})
+    # Act
+    validation_result = argument_validator.validate("label", label, validation_rule)
+    # Assert
+    assert validation_result.is_valid is True
+
+
+@pytest.mark.parametrize("label", [" ", "", "     ", "all", "ALL", "All"])
+def test_validate_label_rejects_empty_strings_or_all(label):
```
Comment (robert-clegg-tessella) [inline piezo_web_app/PiezoWebApp/tests/services/spark_job/validation/argument_validator_test.py]: Please separate out these two scenarios (empty vs all) and check that the validation result message is appropriate.

Comment (robert-clegg-tessella) [reply piezo_web_app/PiezoWebApp/tests/services/spark_job/validation/argument_validator_test.py]: Method name needs updating now that "all" is removed from the parameters ;-)

File: piezo_web_app/PiezoWebApp/src/services/spark_job/spark_job_service.py [hunk 7]
```diff
@@ -70,13 +70,17 @@ def get_job_status(self, job_name, namespace):
             self._logger.error(message)
             raise exception
 
-    def get_jobs(self):
+    def get_jobs(self, label):
+        kwargs = {}
+        if label is not None:
+            kwargs['label_selector'] = ('userLabel=' + label)  # Key must match the key in the manifest populator
```
Comment (robert-clegg-tessella) [inline piezo_web_app/PiezoWebApp/src/services/spark_job/spark_job_service.py]:
``` python
# Key must match the key in the manifest populator
kwargs['label_selector'] = f'userLabel={label}'
```
